### PR TITLE
better explanation about usage and test suite for the object_deserialization

### LINKED ISSUE #320: DeSerializer.object_deserialization does not work
In #318 I noticed, that `Deserializer().object_deserialization()` does fail, while `Deserializer().dwa_deserialization()` works.

I expected `object_deserialization` to be a one-for-all method to deserialize anything (similar to `to_bytes`).

I figured out, why it fails:
https://github.com/PyMoDAQ/PyMoDAQ/blob/9244938e4292b5d06aa02e26b07947fcba6a7ccb/src/pymodaq/utils/tcp_ip/serializer.py#L620-L622

The second line (L621) reads the type, but reading already consumes the bytes from the strings.
Later on the called `deserialization` methods try to read the type again, which fails, as the type has been read and the corresponding bytes are not accessible anymore.

In #318 I propose a fix:
~~~python
    def object_deserialization(self, manual=False):
        """Deserialize an object.

        If you call it manually, you have to set manual to True.
        """
        obj_type = self.string_deserialization()
        hand_over = obj_type if manual else None
        elt = None
        if obj_type == 'scalar':
            elt = self.scalar_deserialization()
        elif obj_type == 'string':
            elt = self.string_deserialization()
        elif obj_type == 'array':
            elt = self.ndarray_deserialization()
        elif obj_type in ('dwa', "DataActuator"):
            elt = self.dwa_deserialization(obj_type=hand_over)
        elif obj_type == "DataToExport":
            elt = self.dte_deserialization(obj_type=hand_over)

    def dte_deserialization(self, obj_type: Optional[str] = None) -> DataToExport:
        """Convert bytes into a DataToExport object

        Convert the first bytes into a DataToExport reading first information about the underlying data

        Returns
        -------
        DataToExport: the decoded DataToExport
        """
        class_name = self.string_deserialization() if obj_type is None else obj_type
~~~

Here, the read obj_type is given to the deserialization method, which does not read the type again.
However, in order to not break code, this possibility is optional and disabled per default.

### PR DESCRIPTION
**Pull request by seb5g**
Related to #323 and closes #320 

Focus on explaining not to use directly the _object_deserialization_ method but only the real object deserialization (dwa, list, scalar, dte, ...)

### CODE REVIEW

**Comment by seb5g** [comment]
I renamed the methods to make them symmetric

I added bytes, dte and ParameterWithPath objects in the type_and_object_serialization/deserialization

The test suite illustrates this symmetry. Is that enough for pyleco?

File: src/pymodaq/utils/parameter/utils.py [hunk 1]
```diff
@@ -1,14 +1,40 @@
...
+            path = get_param_path(parameter)
+        self._path = path
+
+    @property
+    def parameter(self):
+        return self._parameter
+
+    @property
+    def path(self):
+        return self._path
```
Comment (BenediktBurger) [inline src/pymodaq/utils/parameter/utils.py]: If you use type hints for input, why not also for output?
```suggestion
    @property
    def parameter(self) -> Parameter:
        return self._parameter

    @property
    def path(self) -> List[str]:
        return self._path
```

File: src/pymodaq/utils/tcp_ip/serializer.py [hunk 2]
```diff
@@ -617,23 +640,36 @@ def ndarray_deserialization(self) -> np.ndarray:
         ndarray = np.atleast_1d(ndarray)  # remove singleton dimensions
         return ndarray
 
-    def object_deserialization(self):
+    def type_and_object_deserialization(self):
+        """ Deserialize specific objects from their binary representation.
```
Comment (BenediktBurger) [inline src/pymodaq/utils/tcp_ip/serializer.py]: I think it is good to these methods are invers functions, for example:

```suggestion
        """ Deserialize specific objects from their binary representation (inverse of `Serializer.type_and_object_serialization`).
```

Obviously a similar statement could be above and a note in the documentation, that these functions are the entry/exit points of serialization.

File: tests/utils/tcp_ip/serializer_test.py [hunk 3]
```diff
@@ -231,3 +233,95 @@ def test_base_64_de_serialization(get_data: DataToExport):
...
+        serialized = b'\x00\x00\x00\x06string\x00\x00\x00\x0bhello world'
+
+        assert Serializer().type_and_object_serialization(obj) == serialized
+        assert DeSerializer(serialized).type_and_object_deserialization() == obj
+
+    def test_bytes(self):
+        obj = b'hello binary world'
+        serialized = b'\x00\x00\x00\x05bytes\x00\x00\x00\x12hello binary world'
+        assert Serializer().type_and_object_serialization(obj) == serialized
+        assert DeSerializer(serialized).type_and_object_deserialization() == obj
```
Comment (BenediktBurger) [inline tests/utils/tcp_ip/serializer_test.py]: Why don't you parametrize the tests, as they are always the same, only the parameters change?
```suggestion
    @pytest.mark.parametrize("obj, serialized", (
        (True, b'\x00\x00\x00\x06scalar\x00\x00\x00\x03|b1\x00\x00\x00\x01\x01'),
        (10.45, b'\x00\x00\x00\x06scalar\x00\x00\x00\x03<f8\x00\x00\x00\x08fffff\xe6$@'),
        ('hello world', b'\x00\x00\x00\x06string\x00\x00\x00\x0bhello world'),
        (b'hello binary world', b'\x00\x00\x00\x05bytes\x00\x00\x00\x12hello binary world'),
    ))
    def test_serialization(self, obj, serialized):
        assert Serializer().type_and_object_serialization(obj) == serialized
        assert DeSerializer(serialized).type_and_object_deserialization() == obj
```

If you like to add ids (that you know, which one failed), you could add an `ids` list to the parametrize decorator.

File: src/pymodaq/utils/tcp_ip/serializer.py
Comment (BenediktBurger) [inline src/pymodaq/utils/tcp_ip/serializer.py:1]: You can add now `Parameter` to the Serializable list (line 20).

Actually, you could use that list as a type hint for the `type_and_object_serialization` method.

File: tests/utils/tcp_ip/serializer_test.py [hunk 4]
```diff
@@ -231,3 +233,95 @@ def test_base_64_de_serialization(get_data: DataToExport):
...
+
+    def test_array(self):
+        obj = np.array([[0.1, 0.5], [5, 7], [8, 9]])
+        serialized = (b'\x00\x00\x00\x05array\x00\x00\x00\x03<f8\x00\x00\x000\x00\x00\x00\x02\x00'
+                      b'\x00\x00\x03\x00\x00\x00\x02\x9a\x99\x99\x99\x99\x99\xb9?\x00\x00\x00\x00'
+                      b'\x00\x00\xe0?\x00\x00\x00\x00\x00\x00\x14@\x00\x00\x00\x00\x00\x00\x1c@\x00'
+                      b'\x00\x00\x00\x00\x00 @\x00\x00\x00\x00\x00\x00"@')
+
+        assert Serializer().type_and_object_serialization(obj) == serialized
+        assert np.allclose(DeSerializer(serialized).type_and_object_deserialization(), obj)
```
Comment (BenediktBurger) [inline tests/utils/tcp_ip/serializer_test.py:258]: You could include this in the upper, parametrized test as well.

**Review by BenediktBurger — APPROVED** [review]
Looks good.

Thanks for tackling that issue.

I left a few suggestions.

**Comment by BenediktBurger** [comment]
The changes are useful for the leco communication.

Thanks

**Comment by seb5g** [comment]
I merged it in the 4.3.x_dev branch, so you can rebase  yours on it!

**Pull request merged**